### PR TITLE
[RDY] Update Bcash to support native tokens and P2SH32 address type

### PR DESCRIPTION
**Pull request by mainnet-pat**
This pull request upgrades the Bcash to support new BCH network upgrades to include P2SH32 address type support and BCH native tokens (CashTokens) support.

`bchutil` package was [updated](https://github.com/mainnet-pat/bchutil) but referenced as a replacement package. A separate PR for it is out of scope for this review, but can be filed immediately .

<hr>

Blockbook's `bcashparser.go` was updated to support new P2SH32 addresses from `bchutil` and to support token info parsing from output locking bytecode. Extensive tests were added to `bcashparser_tests.go`

New token type structs were put into `types_bcash_type.go`

`rocksdb.go` function `packAddressKey` was updated to strip token info from `AddressDescriptor` for BCH case and store all CashToken utxos under the same key.

`worker.go` functions `getTransactionFromBchainTx` and `GetTransactionFromMempoolTx`  were updated to return BCH specific token data.

`txdetail.html` template and `public.go` were updated to render CashTokens specific data with different token categories being color-coded accordingly as there are more than one different tokens allowed to be interacted with within a single transaction.

`go.sum` updates are unintentional and made by my IDE to compile the project, let me know if we can sort this out with minimal interventions.

<hr>

This PR looks for repository's maintainer feedback regarding data structures and other general remarks which will update its quality.

This PR is marked as WIP, as currently the token metadata (ticker, name, decimals, etc.) lookup and rendering are not implemented, and will be worked on after the submission.

Deployment preview:

Current:

https://bch1.trezor.io/tx/d7ad5f6336cd2c80c86e241ab2deb7b981b7dd9234b68a5b42e304829e482fb2
https://bch1.trezor.io/api/v2/tx/d7ad5f6336cd2c80c86e241ab2deb7b981b7dd9234b68a5b42e304829e482fb2

This PR:
https://blockbook.pat.mn/tx/d7ad5f6336cd2c80c86e241ab2deb7b981b7dd9234b68a5b42e304829e482fb2
https://blockbook.pat.mn/api/v2/tx/d7ad5f6336cd2c80c86e241ab2deb7b981b7dd9234b68a5b42e304829e482fb2

MfG,
pat.

### CODE REVIEW

**Comment by mainnet-pat** [comment]
I am glad to announce that I have finished the work on CashTokens support in blockbook. Token information is nicely parsed from outputs and stored in `TxAddresses` and `AddressBalance` tables and available in API calls. Furthermore, the [BCMR standard](https://github.com/bitjson/chip-bcmr) for token metadata publishing is supported and token category and NFT metadata is stored in RocksDB for fast access from consumers. API and UI were updated to render both CashTokens core info as well as optional metadata (see attached images).

Final sync is being completed and is available at https://blockbook.pat.mn

Please let me know if you have any questions so we can finalize this PR towards the inclusion in the mainline.

### Images
Address page:
image

Transaction info:
image

Token category page:
image

Token nft details page:
image

File: db/rocksdb.go [hunk 1]
```diff
@@ -1199,6 +1301,19 @@ func (d *RocksDB) unpackTxInput(ti *TxInput, buf []byte) int {
...
+			tokenLen, tl := unpackVarint(buf[al:])
+			al += tl
+			if tokenLen > 0 && len(buf[al:]) >= tokenLen {
+				bcashToken, ll, err := bch.UnpackTokenData(buf[al:])
+				if err == nil && ll == tokenLen {
+					al += ll
+					ti.BcashToken = bcashToken
+				}
+			}
+		}
```
Comment (pragmaxim) [inline db/rocksdb.go:1315]: Just fyi, this is not binary compatible and db must be resynced. DB format changes without a version bump will misparse existing DBs.
   - Why it matters: BCH nodes with an existing DB will read old `txAddresses`/`addressBalance` rows with the new token-length fields, leading to silent data corruption or misaligned reads.
   - Suggested fix (force reindex or migrate):
     ```go
     // db/rocksdb.go
     const dbVersion = 8
     ```
     If you want a migration path, add a 7->8 migration in `checkColumns` (similar to 6->7) and gate token fields behind that version.

File: db/rocksdb.go [hunk 2]
```diff
@@ -1101,7 +1176,8 @@ func (d *RocksDB) appendTxOutput(txo *TxOutput, buf []byte, varBuf []byte) []byt
 	return buf
 }
 
-func unpackAddrBalance(buf []byte, txidUnpackedLen int, detail AddressBalanceDetail) (*AddrBalance, error) {
+func (d *RocksDB) unpackAddrBalance(buf []byte, txidUnpackedLen int, detail AddressBalanceDetail) (*AddrBalance, error) {
```
Comment (pragmaxim) [inline db/rocksdb.go:1179]: I think that `unpackAddrBalance` leaks the previous token into subsequent non-token UTXOs.
   - Why it matters: a single token UTXO followed by a normal UTXO will cause the later UTXO to incorrectly carry the token metadata.
   - Suggested fix (reset token per loop):
     ```go
     for len(buf[l:]) >= txidUnpackedLen+3 {
         var bcashToken *bchain.BcashToken
         // ... parse fields ...
         if d.is.IsBCH() && len(buf[l:]) > 0 {
             // parse tokenLen/token data into bcashToken
         }
         u := Utxo{BcashToken: bcashToken, /* ... */ }
         // ...
     }
     ```

File: db/rocksdb.go [hunk 3]
```diff
@@ -1214,7 +1329,21 @@ func (d *RocksDB) unpackTxInput(ti *TxInput, buf []byte) int {
...
+			tokenLen, tl := unpackVarint(buf[al:])
+			al += uint(tl)
+			if tokenLen > 0 && len(buf[l:]) >= tokenLen {
+				bcashToken, ll, err := bch.UnpackTokenData(buf[al:])
+				if err == nil && ll == tokenLen {
+					al += uint(ll)
+					ti.BcashToken = bcashToken
+				}
+			}
+		}
```
Comment (pragmaxim) [inline db/rocksdb.go:1344]: BCH tx input decoding uses the wrong bounds check for token data in the non-extended path.
   - Why it matters: the `len(buf[l:])` check can be true even when there is not enough data remaining, which can misalign the decoder on corrupted/partial rows (or after a failed migration).
   - Suggested fix:
     ```go
     if tokenLen > 0 && len(buf[al:]) >= tokenLen {
         // ...
     }
     ```

File: db/rocksdb.go [hunk 4]
```diff
@@ -646,7 +662,19 @@ func (d *RocksDB) processAddressesBitcoinType(block *bchain.Block, addresses add
...
+			if d.is.IsBCH() {
+				script, _ := hex.DecodeString(output.ScriptPubKey.Hex)
+				var l int
+				bcashToken, l, err = bch.UnpackTokenData(script)
+				if err == nil {
+					addrDesc = script[l:]
+				}
+			} else {
+				addrDesc, err = d.chainParser.GetAddrDescFromVout(output)
+			}
```
Comment (pragmaxim) [inline db/rocksdb.go:677]: BCH output parsing skips P2PK normalization and ignores script decode errors.
   - Why it matters: P2PK outputs are no longer normalized to P2PKH when tokens are present, which changes address indexing behavior compared to legacy BCH handling.
   - Suggested fix (parse token + use parser normalization):
     ```go
     if d.is.IsBCH() {
         script, err := hex.DecodeString(output.ScriptPubKey.Hex)
         if err != nil {
             return err
         }
         bcashToken, _, err = bch.UnpackTokenData(script)
         if err != nil {
             return err
         }
         addrDesc, err = d.chainParser.GetAddrDescFromVout(output)
     }
     ```

File: db/rocksdb_bcash.go [hunk 5]
```diff
@@ -0,0 +1,720 @@
...
+		}
+	}
+
+	if len(metaQueue) > 0 {
+		serializedMetaQueue := make([][]byte, len(metaQueue))
+		for i, mq := range metaQueue {
+			serializedMetaQueue[i] = append(PackBcashTokenMetaQueueKey(mq, make([]byte, 34)), PackBcashTokenMetaQueue(mq, make([]byte, 9))...)
+		}
+		common.BcmrMetaQueueSignal <- serializedMetaQueue
+	}
```
Comment (pragmaxim) [inline db/rocksdb_bcash.go:449]: `processBcashTokens` can block block processing on slow BCMR downloads.
   - Why it matters: the channel has a depth of 1 and the consumer performs synchronous HTTP fetches; this can stall indexing under load or provider latency.
   - Suggested fix (non-blocking send or decouple):
     ```go
     select {
     case common.BcmrMetaQueueSignal <- serializedMetaQueue:
     default:
         glog.Warning("BCMR queue signal dropped; downloader is busy")
     }
     ```

File: server/public.go [hunk 6]
```diff
@@ -665,6 +690,126 @@ func (s *PublicServer) summaryValuesSpan(baseValue float64, secondaryValue float
...
+
+		categoryName := categoryHex
+		if token.Name != "" {
+			categoryName = token.Name
+		}
+
+		tokenCategoryImg := ""
+		if token.Icon != "" {
+			tokenCategoryImg = fmt.Sprintf(`<img onerror="this.style.display='none'" src="%s" alt="Token Icon" width="32px" height="32px"> `, token.Icon)
+		}
```
Comment (pragmaxim) [inline server/public.go:760]: BCMR metadata is rendered into HTML without escaping, enabling XSS via token metadata.
   - Why it matters: `token.Name`, `token.Icon`, and NFT metadata are external inputs that can contain HTML/JS.
   - Suggested fix (escape user-controlled fields or avoid `template.HTML`):
     ```go
     categoryName := template.HTMLEscapeString(token.Name)
     tokenCategoryImg := template.HTMLEscapeString(token.Icon)
     // ... and avoid returning template.HTML for raw concatenated markup
     ```

File: db/rocksdb_test.go [hunk 7]
```diff
@@ -1226,12 +1375,98 @@ func Test_packAddrBalance_unpackAddrBalance(t *testing.T) {
...
+				return
+			}
+			if !reflect.DeepEqual(got1, tt.data) {
+				t.Errorf("unpackTxAddresses() = %+v, want %+v", got1, tt.data)
+			}
+		})
+	}
+}
+
+func Test_packAddrBalance_unpackAddrBalance_Bcash(t *testing.T) {
```
Comment (pragmaxim) [inline db/rocksdb_test.go:1395]: tests do not cover token UTXO followed by non-token UTXO in address balance decoding.
   - Why it matters: this gap would allow the token carry-over bug.
   - Suggested fix: add a test vector where a token UTXO is followed by a non-token UTXO and ensure the latter has `BcashToken == nil`.

**Review by pragmaxim — CHANGES_REQUESTED** [review]
That's a lot of great work here! I made a few comments, cheers

**Comment by pragmaxim** [comment]
@mainnet-pat I made review and you thanked me but did not react on the comments, now there are conflicts since the PR is stale, we were happy to merge this but you did not address the comments

**Comment by cranycrane** [comment]
@mainnet-pat please update the PR if you still want it to be merged

**Comment by cranycrane** [comment]
you should be able to reopen the PR whenever you want
